### ISSUE TITLE
How to send commands to a handler inside a background service?

### ISSUE THREAD

**Issue by MarkLFT** [jbogard/MediatR#679]
I have several micro services that contain a .Net class inherited from BackgroundService.  I need to be able to execute some methods inside the background service. 

Is there a way to send a Mediatr command from elsewhere in the code to a handler defined inside the background service? The methods must be run inside the background service as the commands will send messages over an existing TCP listener connection. 

Thanks in advance for any advice.

**Comment by elementh** [comment]
Indeed, it's possible, not only that but also quite fun.

The only thing you need is a `IServiceScopeFactory _serviceScopeFactory;` in the background service and create an scope every time a new request has to be processed to gather an `ISender` and launch your action.

Something like this:

```csharp
using var scope = _serviceScopeFactory.CreateScope();

var sender = scope.ServiceProvider.GetRequiredService<ISender>();

var result = await sender.Send(new CustomRequest());
```

And then do whatever you want with the response!

This way you get an scoped environment for each request, which is pretty neat and inline with the way MediatR is usually used.

Let me know if this solves your question or if I got something wrong on my understanding of your concerns!

**Comment by MarkLFT** [comment]
Thanks for the reply, but maybe I didn't make myself clear. I will perform the mediatr.Send from somewhere else in the micro service. Inside the background service I want to create a handler to handle a sent message.

**Comment by elementh** [comment]
Oh, then I don't think MediatR is right for your use case, you probably need something else, more like [System.Threading.Channels](https://devblogs.microsoft.com/dotnet/an-introduction-to-system-threading-channels/) and the great helper library [Open.ChannelExtensions](https://github.com/Open-NET-Libraries/Open.ChannelExtensions). 

If you can compromise on sending the data from within the background service or just initiating the connection somewhere else, you could use MediatR no problem.

You can also use a combination of both of them, using MediatR to perform handling of the request and leaving the result inside the channel for your background service to pickup and send over TCP.

**Comment by sanhomealex** [comment]
One of the solutions is to use ConcurrentQueue which will contain messages(commands) that needs to be handled.
BG service has infinite loop that checks for new items (messages) in the queue and passes them to respective handler (better if this handler will run in separate thread or if you prefer `async`, then make sure to set `ConfigureAwait(false)`). 
The queue can be wrapped into a class and the single instance of this class can be injected where needed (including MedaitR handlers) as well as into BG service. 
But please remember that since you are just suppose to place message to the queue, you can not be sure when the particular message being handled, so that this is kind of "send and forget" pattern.

_PS. This is absolutely not related to MediatR at all._

**Comment by eskaufel** [comment]
I am not sure what is going on here. A ConcurrentQueue will only work if both services run inside the same host. Suppose you want to send requests to other services running in different pods/instances. In that case, you need to implement a handler that either directly sends the message over the network or stores the request/notification in an event streaming platform like Kafka, Redis, or Azure Event Hub. Then your consuming applications have a subscription to the topic and transmit them again.

**Comment by sanhomealex** [comment]
Exactly. Concurrent queue will only work inside one process(not the host, cause on one host several processes can run in parallel). And my assumption was that this is what OP meant.
For out of process communications you are right and there are plenty of another techniques/technologies for utilization in these cases.
